### ISSUE TITLE
[TODO] Implement Gradient Tracking Context Manager (NoGradContext)

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#2605]
## Problem

The `NoGradContext` context manager is a stub implementation blocked by Mojo's UnsafePointer parametric mutability limitation.

## Location

- File: `shared/autograd/tape.mojo`
- Lines: 855-879 (NoGradContext struct)
- Lines: 866, 875 (TODO comments)
- Tracked as: TODO(#2400)

**Current stub implementation:**
```mojo
struct NoGradContext:
    """Context manager for disabling gradient tracking (stub).
    
    This is a stub implementation. The full context manager is blocked
    by Mojo's UnsafePointer parametric mutability limitation.
    """
    
    fn __init__(out self):
        """Initialize NoGradContext (stub)"""
        pass
    
    fn __enter__(mut self):
        """Enter no-grad context (stub)"""
        # TODO(#2400): Implement gradient tracking disable
        pass
    
    fn __exit__(mut self):
        """Exit no-grad context (stub)"""
        # TODO(#2400): Implement gradient tracking restore
        pass
```

## Impact

- **Severity**: MEDIUM
- **Functionality**: Cannot disable gradient tracking for inference
- **Performance**: All operations track gradients even when not needed
- **Memory**: Wastes memory storing unused gradients
- **Blocked by**: Mojo compiler limitation (parametric mutability)

## Desired Behavior

```mojo
# In training mode
var loss = model.forward(input)  # Tracks gradients

# In inference mode
with no_grad():
    var output = model.forward(input)  # No gradient tracking
    # Faster, uses less memory
```

## Technical Challenge

The issue is that we need to:
1. Save current gradient tracking state
2. Disable tracking
3. Execute code block
4. Restore previous state

But Mojo's parameter system makes it difficult to pass mutable state through context managers.

## Workaround (Until Mojo Supports This)

Provide function-based API instead of context manager:

```mojo
fn with_no_grad[F: Callable](func: F) -> Result:
    """Execute function without gradient tracking.
    
    Args:
        func: Function to execute
        
    Returns:
        Result of function call
        
    Example:
        var output = with_no_grad(fn() -> ExTensor:
            return model.forward(input)
        )
    """
    var prev_state = Tape.get_grad_enabled()
    Tape.set_grad_enabled(False)
    
    var result = func()
    
    Tape.set_grad_enabled(prev_state)
    return result
```

Or global flag approach:

```mojo
# Global state (less elegant but works)
var GRAD_ENABLED = True

fn set_grad_enabled(enabled: Bool):
    global GRAD_ENABLED
    GRAD_ENABLED = enabled

fn is_grad_enabled() -> Bool:
    return GRAD_ENABLED

# In Tape methods, check flag
fn record_operation(self, ...):
    if not is_grad_enabled():
        return  # Skip recording
    # ... normal recording
```

## Related Issues

- Tracked as #2400 in code
- Blocked by Mojo language limitation
- May be resolved in future Mojo versions
